### PR TITLE
Update for extDB2

### PR DESCRIPTION
**Pull request by Torndeco**
Just updated the code from extDB ->  extDB2 + SQL_CUSTOM

Also added options to enable different extDB2 Protocols i.e steam / rcon / misc (but no code that uses them). Just to make it easier for people to add scripts that might use those features in the future.

Some minor improvements to speed with extDB2, main thing is it alot better handling of losing database connections. 
With original extDB its possible for  it to segfault depending on timing of database connection loses etc..

### CODE REVIEW

**Comment by JoSchaap** [comment]
Thanks alot Torndeco! 

Currently I don't have a active server to test on so I'll have to wait till someone can confirm this is working :)

**Comment by LouDnl** [comment]
I can test this on my test server this evening. No database SQL update needed?

**Comment by Torndeco** [comment]
No database SQL update needed, database scheme is still the same

You just need to copy the sql_custom/a3wasteland.ini  to where you have extDB2 extension at, same as the older db_custom/a3wasteland.ini

If you have some custom SQL Queries you will just  need to update them.
Should only take afew minutes to update them to new format.

**Comment by LouDnl** [comment]
Allright. I will test to tonight on my test server. 


--
This message has been scanned by E.F.A. Project and is believed to be clean.

**Comment by AgentRev** [comment]
Looking good, don't have time to test it myself so I'll will wait on LouD's feedback before merging.

**Comment by LouDnl** [comment]
Allright, configuring files :)

**Comment by LouDnl** [comment]
Torndeco, can you point me to the dll's? Can't seem to find them except uncompiled files?

**Comment by Torndeco** [comment]
https://github.com/Torndeco/extDB2/releases

**Comment by LouDnl** [comment]
Ok so I did the commits on my test server including edits to the main_config.sqf. Copied the a3wasteland.ini to the /extDB/db_custom/ folder. Copied the extDB2.dll tbb.dll and tbbmalloc.dll to arma 3 server root folder. Copied and modified the extdb-conf.ini to point to my database. Now getting this error:

RPT log:
19:06:47 "[extDB2] Startup..."
19:06:47 "[extDB2] ### Database error! [0,"No Config Option Found"]"
19:06:47 "[extDB2] ### ERROR - Startup aborted"

extDB log:
[19:06:47 +02:00] [Thread 1747398523] extDB2: No Config Option Found: A3W
[19:07:02 +02:00] [Thread 1747398523] extDB2: Stopping ...

**Comment by LouDnl** [comment]
Ah I see where I might be able to solve this.
A3W_extDB_ConfigName = "A3W";

**Comment by AgentRev** [comment]
You need to change "[MySQL_Example]" to "[A3W]" in extdb-conf.ini....

**Comment by LouDnl** [comment]
Ah thanks :+1:

**Comment by LouDnl** [comment]
Allright, db_custom needs to be changed to sql_custom.
Getting server crashes now on every startup. They are related to extDB2.dll:
https://gist.github.com/LouDnl/7a46606029f5a1a56711

File: persistence/server/setup/extDB/init.sqf [hunk 1]
```diff
@@ -9,66 +9,86 @@
...
-		diag_log "[extDB] Locked";
-
-		extDB_Database_async = [_setupDir, "async_database.sqf"] call mf_compile;
 		extDB_Misc_async = [_setupDir, "async_misc.sqf"] call mf_compile;
 	};
-
-	extDB_pairsToSQL = [_setupDir, "fn_pairsToSQL.sqf"] call mf_compile;
+	if  (["A3W_extDB_steam"] call isConfigOn) then
+	{
+		extDB_Misc_async = [_setupDir, "async_steam.sqf"] call mf_compile;
```
Comment (AgentRev) [inline persistence/server/setup/extDB/init.sqf]: `extDB_Misc_async` tries to be assigned to both `async_misc.sqf` and `async_steam.sqf` if the 2 options are enabled, is this intended?

**Comment by LouDnl** [comment]
Am not getting it to work. Keep on getting crashes.

**Comment by Soloknight** [comment]
Same here

**Comment by Torndeco** [comment]
Am looking into the issue now, is problem with windows build. 
If you got linux server, it should be fine
I give a shout here when its fixed

**Comment by Soloknight** [comment]
Windows here

File: persistence/server/setup/extDB/init.sqf (hunk 1 above)
Comment (Torndeco) [inline persistence/server/setup/extDB/init.sqf]: Copy / paste mistake have it fixed now

**Comment by Torndeco** [comment]
Should work now, was mistake in a3wasteland.ini

**Comment by LouDnl** [comment]
Seems to be working ok now. Will test more tomorrow.

**Comment by Soloknight** [comment]
SO whats the proper way to install this, I keep getting 
0:12:02 "[extDB2] Startup..."
 0:12:02 "[extDB2] ### Database error! [0,"Database Exception Error"]"
 0:12:02 "[extDB2] ### ERROR - Startup aborted"

**Comment by Torndeco** [comment]
Have a look at extDB2 logs to see more specific info on what the database exception error is for

**Comment by LouDnl** [comment]
Have tested, everything seems to save and load in ok. Loaded in all objects and vehicles from test database right. Warchest works. Bank works. Auto deleting expired vehicles works. Cant seem to find any non working things atm.

**Comment by Soloknight** [comment]
Torndeco, thanks for reply, I have extra stuff to save and I needed to get all the tables right ;-) we will need to fix some script code, but looks good. thanks for your hard work, cheers

**Comment by LouDnl** [comment]
Converting my live servers now.

**Comment by LouDnl** [comment]
Ok my live servers are running extDB2. We'll see how this goes :)
What I do notice is that the extDB2 log file doesn't get written instantly. Even 5 minutes after startup the file is empty. It only writes it sometimes.

**Comment by Torndeco** [comment]
Edit extdb-conf.ini

[Log]
Flush = true

That way it will update the logfile entry for every update

**Comment by LouDnl** [comment]
Thanks

**Comment by Jermin** [comment]
Thank you very much. I need this desperately on my linux box.

**Comment by LouDnl** [comment]
Destroyed vehicles seem to load back in after a restart. Could be that they don't get removed from the database when blown up?

**Comment by Torndeco** [comment]
Fix for deleting objects + vehicles from Database.
Give a shout if you encounter any other issues

**Comment by LouDnl** [comment]
I will, thanks for the fix.

**Comment by AgentRev** [comment]
So, is it stable enough to pull it now?

**Comment by LouDnl** [comment]
I've been running it for a while now and works fine!
Haven't had any problems since latest ini updates.

**Comment by Soloknight** [comment]
running it on 5 servers doing fine

On Sat, Mar 28, 2015 at 4:40 PM, LouDnl <notifications@github.com> wrote:

> I've been running it for a while now and works fine!
> Haven't had any problems since latest ini updates.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/A3Wasteland/ArmA3_Wasteland.Altis/pull/289#issuecomment-87298845>
> .
>



-- 
*Christopher Mitchell*
Blog: MitchellParson.com
Website: StopActionSports.net
Phone: 812-290-3683
Email: Christopher@stopactionsports.net
Connect with me on
<http://www.facebook.com/StopActionSports> *&*
<https://plus.google.com/u/0/115741658571731701625>

**Comment by AgentRev** [comment]
@Torndeco: Just wanted to know, does extDB2 have an option to log all SQL_CUSTOM queries by their INI section name, with all the info passed to and returned by them, like async_database_debug?

Example :

	21:44:59 "DEBUG ASYNC: [1,[1,[[152400]]]]"
	21:44:59 "DEBUG ----- extDB ASync: Complete Time:0.0959473  Input String:getPlayerBankMoney:76561198143978703"

**Comment by AgentRev** [comment]
Also, I see that in "Strip Chars" there are 3 backslashes (\) as opposed to 1 in the older "Bad Chars" option. Are those used for escaping the characters that follow them, or are there just 2 superfluous ones? Do backslashes get stripped or not by default?

**Comment by Torndeco** [comment]
No need for debug info via SQF Code

With normal version, input info is logged when it encounters an error
With debug version all the input info is logged i.e using @extDB2_DEBUG
Its just a different extDB2 dll so you can just swap it with normal extDB2.dll.

-------

The extra slashes in strip Characters is just a copy/paste mistake just remove the extra ones.

Also remove the backslashes from "Strip Custom Chars"  if they are needed etc for updateServerVehicle.

**Comment by Torndeco** [comment]
Bumped requirement for extDB2 to v49

extDB2 was checking for "Strip Custom Input Char" + 
But in documention / a3wasteland.ini i had  "Strip Custom Char"
That is why no-one reported any issues, extDB2 wasn't stripping any characters from $CUSTOM_x$

Fixed now in v49

**Comment by AgentRev** [comment]
Yeah, many of my DB_CUSTOM_V3 queries have their own set of bad chars so as to allow those which are needed for each specific query, while reducing chances of injection.

Do you think you could move the pull request to another branch (e.g. extDB2_beta) so I can merge it to test and tweak before release?

**Comment by Torndeco** [comment]
Got no clue if i can even do that :)
Just make a new branch + do a pull from my fork + you can go make the custom changes etc

**Comment by AgentRev** [comment]
Alright

**Comment by LouDnl** [comment]
I've been running on the latest version of extDB2 and a3wasteland.ini for a day now and all seems fine atm.

**Comment by AgentRev** [comment]
By latest do you mean Torndeco's repo or the extDB2_beta branch?

**Comment by LouDnl** [comment]
extDB2_beta branch

**Comment by AgentRev** [comment]
Alright.

**Comment by LouDnl** [comment]
Not seeing any weird errors, nor am I hearing weird stuff from players.

**Comment by Torndeco** [comment]
Just going to close this to prevent confusion...
Some people are using this instead of the updated extDB2_beta branch

**Comment by AgentRev** [comment]
Good call. If you're wondering why I haven't merged the branch yet, it's because since everybody will have to upgrade extDB to keep the mission working, I'll probably get plenty of requests for help, so I keep it separate until I have more time to handle such requests, when college ends in 6 weeks.

**Pull request closed without merging**